pkg/dataobj/metastore/toc_writer.go

```diff
@@ -96,87 +97,104 @@ func (m *TableOfContentsWriter) WriteEntry(ctx context.Context, dataobjPath stri
 		return err
 	}
 
+	var globalMinTime, globalMaxTime time.Time
+	for _, timeRange := range tenantTimeRanges {
+		if globalMinTime.IsZero() || timeRange.MinTime.Before(globalMinTime) {
+			globalMinTime = timeRange.MinTime
+		}
+		if globalMaxTime.IsZero() || timeRange.MaxTime.After(globalMaxTime) {
+			globalMaxTime = timeRange.MaxTime
+		}
+	}
+
 	// Work our way through the metastore objects window by window, updating & creating them as needed.
 	// Each one handles its own retries in order to keep making progress in the event of a failure.
-	prefix := storagePrefixFor(m.cfg.Storage, m.tenantID)
-	for metastorePath := range iterTableOfContentsPaths(m.tenantID, minTimestamp, maxTimestamp, prefix) {
-		m.backoff.Reset()
-		for m.backoff.Ongoing() {
-			err = m.bucket.GetAndReplace(ctx, metastorePath, func(existing io.ReadCloser) (io.ReadCloser, error) {
-				if existing != nil {
-					defer existing.Close()
-				}
+	for _, timeRange := range tenantTimeRanges {
+		prefix := storagePrefixFor(m.cfg.Storage, string(timeRange.Tenant))
+		for tocPath, tocTimeRange := range iterTableOfContentsPaths(string(timeRange.Tenant), globalMinTime, globalMaxTime, prefix) {
+			m.backoff.Reset()
+			for m.backoff.Ongoing() {
+				err = m.bucket.GetAndReplace(ctx, tocPath, func(existing io.ReadCloser) (io.ReadCloser, error) {
+					if existing != nil {
+						defer existing.Close()
+					}
 
-				m.buf.Reset()
-				m.tocBuilder.Reset()
+					m.buf.Reset()
+					m.tocBuilder.Reset()
 
-				if existing != nil {
-					_, err := io.Copy(m.buf, existing)
-					if err != nil {
-						return nil, errors.Wrap(err, "copying to local buffer")
+					if existing != nil {
+						_, err := io.Copy(m.buf, existing)
+						if err != nil {
+							return nil, errors.Wrap(err, "copying to local buffer")
+						}
 					}
-				}
 
-				if m.buf.Len() > 0 {
-					replayDuration := prometheus.NewTimer(m.metrics.tocReplayTime)
-					object, err := dataobj.FromReaderAt(bytes.NewReader(m.buf.Bytes()), int64(m.buf.Len()))
-					if err != nil {
-						return nil, errors.Wrap(err, "creating object from buffer")
+					if m.buf.Len() > 0 {
+						replayDuration := prometheus.NewTimer(m.metrics.tocReplayTime)
+						object, err := dataobj.FromReaderAt(bytes.NewReader(m.buf.Bytes()), int64(m.buf.Len()))
+						if err != nil {
+							return nil, errors.Wrap(err, "creating object from buffer")
+						}
+						err = m.copyFromExistingToc(ctx, object)
+						if err != nil {
+							return nil, errors.Wrap(err, "reading existing metastore version")
+						}
+						replayDuration.ObserveDuration()
 					}
-					err = m.copyFromExistingToc(ctx, object)
-					if err != nil {
-						return nil, errors.Wrap(err, "reading existing metastore version")
+
+					encodingDuration := prometheus.NewTimer(m.metrics.tocEncodingTime)
+					// Append all the tenant time ranges that overlap with the current Table of Contents window.
+					for _, timeRange := range tenantTimeRanges {
```

Comment:
I'm confused here. `tenantTimeRanges` is multi-tenant right? But then we read from a tenant prefixed path, and then add the multi-tenant time ranges to it? That doesn't seem correct. Have I misunderstood this code?